intermediate_source/torch_export_tutorial.py

```diff
@@ -0,0 +1,462 @@
+# -*- coding: utf-8 -*-
+
+"""
+torch.export Tutorial
+================
+**Author:** William Wen, Zhengxu Chen
+"""
+
+######################################################################
+# ``torch.export`` is the PyTorch 2.0 way to export PyTorch models intended
+# to be run on high performance environments.
+#
+# ``torch.export`` is built using the components of ``torch.compile``,
+# so it may be helpful to familiarize yourself with ``torch.compile``.
+# For an introduction to ``torch.compile``, see the ` ``torch.compile`` tutorial <https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html>`__.
+#
+# This tutorial focuses on using ``torch.export`` to extract
+# `ExportedProgram`s (i.e. single-graph representations) from PyTorch programs.
+#
+# **Contents**
+#
+# - Exporting a PyTorch model using ``torch.export``
+# - Comparison to ``torch.compile``
+# - Control Flow Ops
+# - Constraints
+# - Custom Ops
+# - ExportDB
+# - Conclusion
+
+######################################################################
+# Exporting a PyTorch model using ``torch.export``
+# ------------------------------------------------
+#
+# ``torch.export`` takes in a callable (including ``torch.nn.Module``s),
+# a tuple of positional arguments, and optionally (not shown in the example below),
+# a dictionary of keyword arguments.
+
+import torch
+from torch.export import export
+
+class MyModule(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.lin = torch.nn.Linear(100, 10)
+
+    def forward(self, x, y):
+        return torch.nn.functional.relu(self.lin(x + y), inplace=True)
+
+mod = MyModule()
+exported_mod = export(mod, (torch.randn(8, 100), torch.randn(8, 100)))
+print(type(exported_mod))
+
+######################################################################
+# ``torch.export`` returns an ``ExportedProgram``, which is not a ``torch.nn.Module``,
+# but can still be ran as a function:
+
+print(exported_mod(torch.randn(8, 100), torch.randn(8, 100)))
+
+######################################################################
+# ``ExportedProgram`` has some attributes that are of interest.
+# The ``graph`` attribute is an FX graph traced from the function we exported,
+# that is, the computation graph of all PyTorch operations.
+# The FX graph has some important properties:
+# - The operations are "ATen-level" operations.
+# - The graph is "functionalized", meaning that no operations are mutations.
```

Comment:
Is this configurable? iirc you don't have to get an aten graph, right?

Reply:
(I think) current state is that you only get an aten graph, and then (eventually) there'll also be a way to get out a "pre_dispatcher/pre_dispatch" aten graph.

`torch._dynamo.export()` can technically do more and return a torch graph, but there was some discussion around `torch.export()` being more opinionated and wanting to be able to give a guarantee that the graph is always functional (which means torch IR is a no-go).